providers/providers.go

```diff
@@ -83,36 +83,49 @@ func (pm *ProviderManager) Process() goprocess.Process {
 	return pm.proc
 }
 
-func (pm *ProviderManager) providersForKey(k []byte) ([]peer.ID, error) {
-	pset, err := pm.getProvSet(k)
+// Yields the PeerIDs of the providers from local (cache or datastore) for a given key
+func (pm *ProviderManager) getProvidersForKey(k []byte) ([]peer.ID, error) {
+	pset, err := pm.getProviderSetForKey(k)
 	if err != nil {
 		return nil, err
 	}
 	return pset.providers, nil
 }
 
-func (pm *ProviderManager) getProvSet(k []byte) (*providerSet, error) {
-	cached, ok := pm.providers.Get(string(k))
+// Yields a ProvidersSet from local (cache or datastore). Checks cache first
+func (pm *ProviderManager) getProviderSetForKey(k []byte) (*providerSet, error) {
+	cached, ok := pm.cache.Get(string(k))
 	if ok {
 		return cached.(*providerSet), nil
 	}
 
-	pset, err := loadProvSet(pm.dstore, k)
+	pset, err := loadProviderSet(pm.dstore, k)
 	if err != nil {
 		return nil, err
 	}
 
 	if len(pset.providers) > 0 {
-		pm.providers.Add(string(k), pset)
+		pm.cache.Add(string(k), pset)
 	}
 
 	return pset, nil
 }
 
-func loadProvSet(dstore ds.Datastore, k []byte) (*providerSet, error) {
-	res, err := dstore.Query(dsq.Query{Prefix: mkProvKey(k)})
-	if err != nil {
-		return nil, err
+// Retried provider records from datastore
+func loadProviderSet(dstore ds.Datastore, k []byte) (*providerSet, error) {
```

Comment:
This could go very wrong in very bad ways. If someone provides an empty key, I'd expect an error. Instead, this will likely hang for a while and allocate a _ton_ of memory. If we're not careful, someone on the network could trigger this issue by sending us a request for a nil/empty provider key.

Reply:
Also, was this the bug? This is not a bug, this function was never designed to enumerate all provider records.

Reply:
If you actually need to be able to list all providers, we'll need a new function that can stream them back on a channel.

Reply:
Not the bug. I started by wanting to get a GetAllProviderRecords and found that cache was not being updated.

I'll separate the feature from the bug fix.

Good call on sending empty key. I'll just create a separate func then